lib/dialects/abstract/query.js

```diff
@@ -7,6 +7,46 @@ const Dot = require('dottie');
 const deprecations = require('../../utils/deprecations');
 const uuid = require('uuid/v4');
 
+const stringify = obj => obj instanceof Buffer ? obj.toString('hex') : obj;
+
+// Calculates the array prefix of a key (['User', 'Results'] for 'User.Results.id')
+const makeDoublePrefixMemoizer = () => {
+  const keyPrefixMemo = {};
+  return str => {
+    // We use a double memo and keyPrefixString so that different keys with the same prefix will receive the same array instead of differnet arrays with equal values
+    if (!Object.prototype.hasOwnProperty.call(keyPrefixMemo, str)) {
```

Comment:
For better readability

```Reflect.apply(Object.hasOwnProperty, keyPrefixMemo, [str]```

Reply:
This feels rather subjective, if anything I'd like to generate all usage of `Object.prototype.hasOwn...` to a util called `Utils.hasOwnProperty(target, str)` but this is out of scope of this PR.
The temporary array is also likely to incur another performance hit.

I'll do some benchmarking around this, I never really worked with the `Reflect` library, it feels like a clone of `Object`.

Reply:
Naive benchmark:
```js
'use strict';

const obj = {
};

function randIndex() {
  return Math.floor(Math.random() * 100);
}

for (let i = 0; i < 100;++i) {
  obj[randIndex()] = Math.random();
}

let out;
console.time('Reflect.apply');
for (let i = 0; i < 0xffffff; i++) {
  out = Reflect.apply(Object.hasOwnProperty, obj, [randIndex()]);
}
console.timeEnd('Reflect.apply');

console.time('Object.prototype.hasOwnProperty.call');
for (let i = 0; i < 0xffffff; i++) {
  out = Object.prototype.hasOwnProperty.call(obj, randIndex());
}
console.timeEnd('Object.prototype.hasOwnProperty.call');

const bound = Object.prototype.hasOwnProperty;
function hasOwnProp(obj, prop) {
  return bound.call(obj, prop);
}

console.time('hasOwnProp');
for (let i = 0; i < 0xffffff; i++) {
  out = hasOwnProp(obj, randIndex());
}
console.timeEnd('hasOwnProp');

console.log(out);
```

```
Reflect.apply: 547.062ms
Object.prototype.hasOwnProperty.call: 353.256ms
hasOwnProp: 343.112ms

Reflect.apply: 544.96ms
Object.prototype.hasOwnProperty.call: 353.499ms
hasOwnProp: 357.777ms

Reflect.apply: 547.306ms
Object.prototype.hasOwnProperty.call: 354.269ms
hasOwnProp: 350.131ms
```

Reflect is quite a bit slower than the Object variant, the wrapped (tidy) version is sometimes faster, sometimes slower but pretty much the same speed on average, I will likely file a PR to replace those later.